### PR TITLE
AutoMM Detection Tutorial (VOC to COCO, Finetune)

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
- [Tutorial] Fast Finetune on VOC in COCO format (YoloV3, time = 100.14 seconds)
- [Tutorial] High Performance Finetune on VOC in COCO format  (VFNet, mAP50 = 0.92)
- [Tutorial] Convert VOC Format to COCO Format
- Fix previous hard code while reading MMDet config
- Raise error for MaskXXX models.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+and also provide a `sample_data_path` for the predictor to infer the catgories of the dataset.
+Here we provide the `training_path`, and it also works using any other split of this dataset.
+
+```python
+predictor = MultiModalPredictor(
+    hyperparameters={
+        "model.mmdet_image.checkpoint_name": checkpoint_name,
+        "env.num_gpus": num_gpus,
+    },
+    pipeline="object_detection",
```
Comment (bryanyzhu) [inline docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:43]: In the text, you mentioned problem_type, but here is still pipeline, we need to fix this later (probably not now).

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    sample_data_path=train_path,
+)
+```
+
+If no data sample is available at this point, you can also create the MultiModalPredictor by manually input the classes:
```
Comment (bryanyzhu) [inline docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:48]: What do you mean by no data sample is available? Like no json? Because if users follow our tutorial, they will have train/text json by now. We need more context here to understand when this use case happen.

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+num_gpus = -1  # use all GPUs
+
+train_path = "./VOCdevkit/VOCCOCO/voc12_train.json"
+test_path = "./VOCdevkit/VOCCOCO/voc07_test.json"
+```
+
+We create the MultiModalPredictor with selected checkpoint name and number of GPUs.
+We need to specify the problem_type to `"object_detection"`,
+and also provide a `sample_data_path` for the predictor to infer the catgories of the dataset.
+Here we provide the `training_path`, and it also works using any other split of this dataset.
```
Comment (bryanyzhu) [inline docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:35]: In the code, I think you wrote `train_path`, but here is `training_path`, will that make a difference?

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        "env.num_gpus": num_gpus,
+    },
+    pipeline="object_detection",
+    classes=voc_classes,
+)
+```
+
+We set the learning rate to be `1e-4`.
+Note that we use a two-stage learning rate option during finetuning by default,
+and the model head will have 100x learning rate.
```
Comment (bryanyzhu) [inline docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:64]: Need 1 or 2 sentence here to describe why we want to do this two-stage thing. Users might not have the knowledge of detection, or two-stage finetuning technique.

File: docs/tutorials/multimodal/object_detection/detection_high_performance_finetune_coco.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+ Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.792                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.569                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.811                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.827                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.603                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.754                                                                               
+ Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.866  
+```
+
+Under this high performance finetune setting, it took 5 hours but reached `mAP50 = 0.932` on VOC!
```
Comment (bryanyzhu) [inline docs/tutorials/multimodal/object_detection/detection_high_performance_finetune_coco.md:156]: We can write one sentence here to compare with performance using yolov3_mobilenetv2. Let users have a clear understanding what is the accuracy and speed tradeoff is.

File: examples/automm/object_detection/finetune_coco_format.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        --checkpoint_name yolov3_mobilenetv2_320_300e_coco \
+        --lr <learning_rate> \
+        --epochs <epochs>
+
+An example to finetune an MMDetection model on VOC:
+    First, use this script to convert the VOC dataset to COCO format:
+    https://github.com/open-mmlab/mmdetection/blob/9d3e162459590eee4cfc891218dfbb5878378842/tools/dataset_converters/pascal_voc.py
+    Then, run:
+    python finetune_coco_format.py \
+        --train_path /media/data/datasets/voc/VOCdevkit/VOCCOCO/voc07_trainval.json \
```
Comment (bryanyzhu) [inline examples/automm/object_detection/finetune_coco_format.py:17]: Use relative path or placeholder

File: multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py [hunk 7]
```diff
@@ -88,6 +88,13 @@ def _get_map_input(self, pred_results, sample):
             boxes = []
             scores = []
             labels = []
+            if isinstance(img_result[0], list):
+                # has one additional dimention with 2 outputs: img_result=img_result[0], mask_result=img_result[1]
+                raise NotImplementedError(
+                    "Do not support training for models with masks like mask r-cnn, "
```
Comment (bryanyzhu) [inline multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py:94]: So for all models with mask branch, we can't train the model? Or we just don't have results for mask rcnn like model?

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 1 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:43]: Yes, I'll replace all the `pipeline` to problem type once Xingjian's PR is merged

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 2 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:48]: A possible use case is online finetune and inference. And thus there is no data available while initialize the predictor.

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 3 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:35]: Oh that's a typo, will be fixed in next pr.

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 4 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:64]: Added "Using a two-stage learning rate with high learning rate only on head layers makes
the model converge faster during finetuning. It usually gives better performance as well,
especially on small datasets with hundreds or thousands of images.".

File: examples/automm/object_detection/finetune_coco_format.py (hunk 6 above)
Comment (FANGAreNotGnu) [reply examples/automm/object_detection/finetune_coco_format.py:17]: Will change this after Yongxing's PR is merged. There are some VOC related changes in that PR.

File: docs/tutorials/multimodal/object_detection/detection_high_performance_finetune_coco.md (hunk 5 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_high_performance_finetune_coco.md:159]: Done, also added the comparison in yolov3 finetune tutorial

File: multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py (hunk 7 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/optimization/lit_mmdet.py:94]: Currently we can't train mask-rcnn, cascade mask-rcnn, and htc.

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 2 above)
Comment (bryanyzhu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:48]: Do you think we need a new tutorial for online case? or just put it here?

**Review by bryanyzhu — APPROVED** [review]
Thank you for the great tutorials.

File: docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md (hunk 2 above)
Comment (FANGAreNotGnu) [reply docs/tutorials/multimodal/object_detection/detection_fast_finetune_coco.md:48]: I reconsidered this and think we can remove this paragraph. It makes sense that we ask the user to initialize the predictor only when they have sample data.

**Review by sxjscience — APPROVED** [review]
LGTM
